Added linalg.eig, linalg.eigvals

IvanYashchuk opened the pull request:
This PR adds `torch.linalg.eig`, and `torch.linalg.eigvals` for NumPy compatibility.

MAGMA uses a hybrid CPU-GPU algorithm and doesn't have a GPU interface for the non-symmetric eigendecomposition. It means that it forces us to transfer inputs living in GPU memory to CPU first before calling MAGMA, and then transfer results from MAGMA to CPU. That is rather slow for smaller matrices and MAGMA is faster than CPU path only for matrices larger than 3000x3000.
Unfortunately, there is no cuSOLVER function for this operation.

Autograd support for `torch.linalg.eig` will be added in a follow-up PR.

Ref #42666

IvanYashchuk committed:
Added linalg.eig

IvanYashchuk committed:
Added linalg.eigvals

IvanYashchuk committed:
flake8

nikitaved commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  auto real_vectors_data = real_vectors.data_ptr<scalar_t>();
+  auto values_data = complex_values.data_ptr<c10::complex<scalar_t>>();
+
+  for (auto b = decltype(batch_size){0}; b < batch_size; b++) {
+    scalar_t* vecs = &real_vectors_data[b * matrix_stride];
+    c10::complex<scalar_t>* res = &result_data[b * matrix_stride];
+    c10::complex<scalar_t>* vals = &values_data[b * n];
+    for (auto j = decltype(n){0}; j < n; j++) {
+      if (vals[j].imag() == 0.0) {  // eigenvalue is real, then v(j) = VR(:,j)
+        for (auto i = decltype(n){0}; i < n; i++) {
+          res[j * n + i] = c10::complex<scalar_t>(vecs[j * n + i], 0);
+        }
+      } else {
+        for (auto i = decltype(n){0}; i < n; i++) {
+          res[j * n + i] = c10::complex<scalar_t>(vecs[j * n + i],  vecs[(j+1) * n + i]);      // v(j)   = VR(:,j) + i*VR(:,j+1)
+          res[(j+1) * n + i] = c10::complex<scalar_t>(vecs[j * n + i], -vecs[(j+1) * n + i]);  // v(j+1) = VR(:,j) - i*VR(:,j+1)
+        }
+        j++;
+      }
+    }
+  }
+}
```
You can make this code device-independent with eigenvalue synchronization if on CUDA by doing something along these lines:
```
res_3d = res.view({-1, res.size(-2), res.size(-1)});
values_2d = values.view({-1, values.size(-1)});
real_vectors_3d = real_vectors.view({-1, real_vectors.size(-2), real_vectors.size(-1)});
for (b = ....) {
   res = res_3d.select(0, b);
   values = values_2d.select(0, b);
   real_vectors = real_vectors_3d.select(0, b):
   for (j = ....) {
       res.select(-1, j) = real_vectors.select(-1, j) + i * real_vectors.select(-1, j + 1);
   }
}
```

IvanYashchuk replied:
Yes, I have the following prototype working
```
  result.copy_(real_vectors);
  auto conjugate_pair_indices = (at::imag(complex_values) > 0).nonzero().split(1, 1);
  auto shifted_indices = conjugate_pair_indices;
  shifted_indices.back() = shifted_indices.back() + 1;
  auto result_imag = at::imag(result);
  auto indices = at::native::toListOfOptionalTensors(conjugate_pair_indices);
  auto shifted_indices_ = at::native::toListOfOptionalTensors(shifted_indices);
  result_imag.transpose(-2, -1).index_put_(std::move(indices), real_vectors.transpose(-2, -1).index(std::move(shifted_indices_)));
  result.transpose(-2, -1).index_put_(std::move(shifted_indices_), result.transpose(-2, -1).index(indices).conj());
  return result;
  ```
  It didn't give any speedup when I checked, maybe some things could be made faster.
  But here we always have contiguous inputs that are always on the CPU and no computations just moving the memory around, so the compiler should make it efficient.

nikitaved replied:
Something like this could probably work:
```
auto idx_pure_real = (at::imag(values_2d) == 0.0).nonzero().squeeze(-1) // returns 2d tensor, but we need 1d
res_3d.index_put_({idx_pure_real}, at::index_select(real_vectors_2d, 0, idx_pure_real));
```

nikitaved replied:
I also suspect that you can do this without even the per-matrix loop over `for(b = ...)`.

IvanYashchuk committed:
Make shorter line

mruberry commented:
Hey @IvanYashchuk! As I'm catching up on PRs: what's the status of this one? Is it ready for review?

IvanYashchuk commented:
> Hey @IvanYashchuk! As I'm catching up on PRs: what's the status of this one? Is it ready for review?

Yes, it is. The only thing I might need to change is the `linalg_eig_make_complex_eigenvectors_impl` function I still haven't profiled this part properly. I might replace it with "at:: vectorized" code https://github.com/pytorch/pytorch/pull/52491#discussion_r579022208.
Everything else is certainly ready for review.

mruberry commented:
> Unfortunately, there is no cuSOLVER function for this operation.

@xwang233 -- maybe another request for the cuSOLVER team? We'll need some strategy for replacing MAGMA here.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+
+static Tensor& linalg_eig_make_complex_eigenvectors(Tensor& complex_vectors, const Tensor& complex_values, const Tensor& real_vectors) {
+  // These asserts make explicit the requirements on tensors for 'linalg_eig_make_complex_eigenvectors_impl'
+  TORCH_INTERNAL_ASSERT(complex_vectors.device() == at::kCPU);
```
Great asserts.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+DEFINE_DISPATCH(linalg_eig_stub);
+
+std::tuple<Tensor&, Tensor&> linalg_eig_out_info(const Tensor& input, Tensor& values, Tensor& vectors, Tensor& infos, bool compute_eigenvectors) {
+  // MAGMA doesn't have GPU interface for GEEV routine, it requires inputs to be on CPU
```
Nice comments.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  }
+
+  // if 'values' has no elements we can modify it
+  auto values_shape = IntArrayRef(input.sizes().data(), input.dim()-1);  // input.shape[:-1]
```
How does this interact with the internal asserts later?

If values can come from users, then shouldn't the later internal asserts be torch checks? If values only comes internally, shouldn't its properties just be asserted?

IvanYashchuk replied:
This function has two modes of operation:

- if passed "out" tensors are empty then they can be resized, this makes calling this function from `linalg_eig` and `linalg_eig_out` simpler.
- if non-empty tensors are passed they must be in lapack compatible format: column-major memory layout, contiguous, and of correct shape.

Possible resizing of passed by user tensors happen outside of this function in `linalg_eig_out`. This avoids recursion and separates the scope of functions.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    if (values.is_complex()) {
+      values = at::complex_out(values, real_values, imag_values);
+    } else {
+      TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvalues is non-zero, can't safely cast eigenvalues to non-complex dtype.")
```
"linalg_eig" -> "torch.linalg.eig()". Let's start preferring the user-facing names in the error messages.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    auto imag_values = real_imag_values.slice(/*dim=*/-1, /*start=*/input.size(-1));
+
+    // if the imaginary part is zero we don't need to do anything
+    bool is_zero_imag = at::all(imag_values == 0.0).item().toBool();
```
This would sync the GPU and CPU even if other parts of the algorithm didn't.

IvanYashchuk replied:
It would, if `imag_values` lived on GPU, but it's always stored on the CPU side currently.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  // if 'input' is complex then use 'vectors' directly else maybe create a temporary to hold real vectors
+  // and then use linalg_eig_make_complex_eigenvectors
+  Tensor maybe_complex_vectors = vectors;
+  if (!input.is_complex()) {
```
This is a lot of work for inputs that aren't complex.

Would you elaborate on what's going on with floating point vs complex inputs? Does it make sense to cast all inputs to complex?

IvanYashchuk replied:
Casting all inputs to complex dtype would make LAPACK perform unnecessarily worse. This post-processing is required for inputs that aren't complex just because LAPACK gives output in compressed form using only real-valued arrays.
I will profile the implementation to show this is actually not a lot of work computationally.

IvanYashchuk replied:
For a matrix of size 4096x4096 `linalg_eig_make_complex_eigenvectors` takes ~0.16% of the total execution time of `linalg_eig_out_info`, ~99.84% is spent in `linalg_eig_stub`.
For a matrix of size 256x256 `linalg_eig_make_complex_eigenvectors` takes ~0.6%.

mruberry replied:
OK.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+    is_batched_column_major = vectors.transpose(-2, -1).is_contiguous();
+  }
+
+  // for NumPy compatibility we can't determine the dtype of eigenvalues before running the computation
```
Let's talk more about this.

So in NumPy if the input is real-valued then the dtype of the returned tensor is real if the eigenvalues are real-valued and complex if the eigenvalues have non-zero imaginary parts?

How is that actually implemented in NumPy?

Should we mimic this behavior? PyTorch tends to prefer static shape propagation. That is, we won't be able to write shape propagation logic for this operation if its output dtype is value-dependent. Note that JAX does NOT implement this NumPy behavior either (they're very dependent on static shape propagation):

```
np.linalg.eig(np.diag((1, 2, 3)))
: (array([1., 2., 3.]), array([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]))

jnp.linalg.eig(np.diag((1, 2, 3)))
: [DeviceArray([1.+0.j, 2.+0.j, 3.+0.j], dtype=complex64),
 DeviceArray([[1.+0.j, 0.+0.j, 0.+0.j],
              [0.+0.j, 1.+0.j, 0.+0.j],
              [0.+0.j, 0.+0.j, 1.+0.j]], dtype=complex64)]
```

cc @rgommers, this is another interesting example of shape propagation logic considerations that the array-api probably wants to consider

IvanYashchuk replied:
> So in NumPy if the input is real-valued then the dtype of the returned tensor is real if the eigenvalues are real-valued and complex if the eigenvalues have non-zero imaginary parts?
How is that actually implemented in NumPy?

That's right. NumPy checks whether all values of the imaginary part of eigenvalues are equal to 0.0 with no tolerance, if that's the case casts output to reals.
https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/linalg/linalg.py#L1325

In Julia, it's the same behavior (all the imaginary values of eigenvalues are also compared to be == 0)
https://github.com/JuliaLang/julia/blob/788b2c77c10c2160f4794a4d4b6b81a95a90940c/stdlib/LinearAlgebra/src/eigen.jl#L157

mruberry replied:
PyTorch's correct behavior would be easier to determine if we were further along in our program acquisition work. For now I would suggest we follow JAX and warn that this behavior may change in the future, however. That is, I suspect our method of program acquisition will require shape propagation logic to be value-independent.

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  vectors_tmp_needed |= !vectors_expected_type;
+  // we will allocate a temporary tensor and do the copy
+
+  // "out=" variants shouldn't allow result and input to be on different devices
```
https://github.com/pytorch/pytorch/pull/53259 will add checks for this to all OpInfos

IvanYashchuk replied:
This function is special. MAGMA takes CPU inputs and return CPU outputs and internally calls GPU for some parts of the computation managing CPU-GPU data transfers. If the passed tensors to out= live on CPU while the input is on GPU, then it's actually a beneficial situation involving fewer explicit CPU-GPU copies. Should we be consistent and disallow out= and input to be on different devices?

mruberry replied:
Yes. I appreciate your reasoning and you're completely right, but supporting these little oddities has some cost and in the future we may be able to switch to a different implementation that is actually run on the GPU, and then our oddity will become a liability.

mruberry commented on test/test_linalg.py:
```diff
@@ -1533,6 +1533,296 @@ def test_eig_out(self, device, dtype):
+                # for symmetric real-valued inputs eigenvalues and eigenvectors are also real-valued
+                a = random_symmetric_matrix(shape[-1], *shape[:-2], dtype=dtype, device=device)
+            else:
+                a = torch.randn(*shape, dtype=dtype, device=device)
```
Use make_tensor

mruberry commented on test/test_linalg.py:
```diff
@@ -1533,6 +1533,296 @@ def test_eig_out(self, device, dtype):
+            expected = (np.take_along_axis(expected[0], ind, axis=-1), np.take_along_axis(expected[1], ind[:, None], axis=-1))
+
+            # sort PyTorch output
+            # torch.argsort doesn't work well with complex inputs, NumPy sorting on CPU is used instead
```
Would you elaborate on "doesn't work well with..."?

IvanYashchuk replied:
It just doesn't work. The comment shouldn't have the "well" word.

mruberry commented on test/test_linalg.py:
```diff
@@ -1533,6 +1533,296 @@ def test_eig_out(self, device, dtype):
+
+            # sort PyTorch output
+            # torch.argsort doesn't work well with complex inputs, NumPy sorting on CPU is used instead
+            # RuntimeError: _th_sort not supported on CUDAType for ComplexDouble
```
This is correct. We don't support sorting complex numbers because they're not part of any totally ordered field. So how to sort them is ???. Supporting explicit lexical sorting would be nice, though.

IvanYashchuk replied:
This is true, but still, we need to sort the values based on some criteria. In this case, we just rely on `np.argsort` to bring PyTorch and NumPy results to something comparable.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  // for NumPy compatibility we can't determine the dtype of eigenvalues before running the computation
+  // for real-valued inputs NumPy can return either real- or complex-valued output, we will create values_tmp of complex type
+  // for complex-valued inputs we can use values' storage directly if dtype matches
+  // bool values_input_same_type = input.is_complex() ? (values.scalar_type() == input.scalar_type()) : false;
```
Remove this line.

mruberry commented on test/test_linalg.py:
```diff
@@ -1533,6 +1533,296 @@ def test_eig_out(self, device, dtype):
+    @skipCUDAIfNoMagma
+    @skipCUDAIfRocm
+    @dtypes(*floating_and_complex_types())
+    def test_eig_compare_backends(self, device, dtype):
```
Would you elaborate on this test vs. comparing both the CPU and GPU with the NumPy reference?

If we want to compare CPU vs. GPU we should probably mark this test @onlyCPU and short-circuit if CUDA is not available.

IvanYashchuk replied:
The above `test_eig_numpy` only compares float64 and complex128 dtype with NumPy. Results for float32 and complex64 differ a lot from NumPy (because internally NumPy casts inputs to double precision) and I thought it would be good to check CPU vs GPU whether we get the same values.

In addition this test checks the out= variant, but this could probably be handled with OpInfo now.

mruberry replied:
out= should be handled by opinfo testing now.

I appreciate your thinking on float32 and complex64. SGTM.

mruberry commented on test/test_linalg.py:
```diff
@@ -1533,6 +1533,296 @@ def test_eig_out(self, device, dtype):
+                self.assertEqual(expected[0].to(complex_dtype), out0)
+                self.assertEqual(expected[1].to(complex_dtype), out1)
+
+                # check non-contiguous out
```
Luckily these tests will be performed by the out= PR I mentioned so we won't have to keep writing them!

mruberry replied:
Now that the out= PR is merged yes, I think we can remove these.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -235,6 +235,102 @@
+
+.. note:: The eigenvalues are computed using LAPACK's and MAGMA's `geev` routine.
+
+.. note:: See :func:`torch.linalg.eig` for a related function that computes both eigenvalues and eigenvectors.
```
and is differentiable

IvanYashchuk replied:
Let's add it once `torch.linalg.eig` is actually differentiable.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1744,6 +1744,33 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           op=torch.linalg.eig,
+           dtypes=floating_and_complex_types(),
+           test_inplace_grad=False,
+           # TODO: TypeError: empty_like(): argument 'input' (position 1) must be Tensor,
```
This should be fixed with the linked out= PR

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1744,6 +1744,33 @@ def sample_inputs_masked_select(op_info, device, dtype, requires_grad):
+           decorators=[skipCUDAIfNoMagma, skipCUDAIfRocm, skipCPUIfNoLapack],
+           skips=(
+               # TODO: RuntimeError: derivative for linalg_eig is not implemented
+               SkipInfo('TestGradients'),
```
The "supports_autograd" property should handle these skips now

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -235,6 +235,102 @@
+          different device types. Usually the difference is only in the sign of the eigenvector.
+
+.. note:: See :func:`torch.linalg.eigvals` for a related function that computes only eigenvalues.
+          However, that function is not differentiable.
```
But isn't this function also not differentiable? The docs should probably reflect that, even if the state is only temporary

IvanYashchuk replied:
Added a warning about that.

mruberry reviewed (commented):
Hey @IvanYashchuk! This is a pretty awesome PR. I made some inline comments but also have several questions about the design that I look forward to hearing your thoughts on.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Change linalg_eig -> torch.linalg.eig

IvanYashchuk committed:
Remove commented code

IvanYashchuk committed:
Added checks for the same device of out and input

IvanYashchuk committed:
randn -> make_tensor

IvanYashchuk committed:
Fix comment

IvanYashchuk committed:
Make test_eig_compare_backends run only on cuda

IvanYashchuk committed:
Added a warning that differentiation of eig is not implemented yet

IvanYashchuk committed:
Remove supports_tensor_out

IvanYashchuk committed:
Use supports_autograd

IvanYashchuk committed:
Trim trailing whitespace

nikitaved commented:
@IvanYashchuk , the `eig_backward` got just merged. Could you please test whether it is compatible with `linalg.eig`?

IvanYashchuk committed:
Merge branch 'master' into linalg-eig

IvanYashchuk committed:
Fix the result to be always complex-valued

even if the imaginary part of the eigenvalues is zero.
This change is incompatible with NumPy that casts the result to real
dtype.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Removed out= test cases; they're handled with OpInfo tests

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Changed TORCH_INTERNAL_ASSERT -> TORCH_INTERNAL_ASSERT_DEBUG_ONLY

IvanYashchuk committed:
Fix merge error

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

anjali411 commented:
@IvanYashchuk is this PR ready for a final review?

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Removed test_inplace_grad=False

IvanYashchuk commented:
There was a merge conflict failing tests, now it is resolved. I think I've addressed all review comments. It is ready for another round of review @anjali411, @mruberry. 

An open question still whether for CUDA inputs we should internally transfer them to CPU to compute the result. As the performance results suggest it is faster for matrices up to 2048x2048 https://github.com/pytorch/pytorch/pull/52491#issuecomment-795685687.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into linalg-eig

IvanYashchuk committed:
Call CPU path for inputs less than 2048x2048

IvanYashchuk commented:
@mruberry I resolved the merge conflicts and added the dispatch to CPU for CUDA inputs smaller than 2048x2048.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8557,6 +8557,28 @@
+  python_module: linalg
+  variants: function
+  dispatch:
+    CPU, CUDA: linalg_eig
```
Functional variant here and below redispatch, so CompositeExplicitAutograd would also have worked too. But even better, consider making this structured kernel https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

mruberry replied:
Structured kernel support seems like a good follow-up PR.

mruberry reviewed (approved):
Cool! Let's get this landed now so we can stack future changes on it.
